### PR TITLE
Feature uml diagrams

### PR DESCRIPTION
**Pull request by itziakos**
This PR add the UML diagrams for the current CUDS api description

in more detail:

- The diagrams are described using plantuml in plain text inside the source/uml folder
- The plantuml.jar file is added and a new `make` target `uml` is added to build the png files for the documentation as necessary.
- The README is updated with instraction on setting up the plantuml build enviroment
- A new page is was added for the CUDS API description using the uml diagrams
- Minor cleanup has took place to bridge differences between the abstract classes and the native implementation (mainly signature variables).

### CODE REVIEW

**Comment by itziakos** [comment]
fixes #95

**Comment by itziakos** [comment]
@roigcarlo  and @nathanfranklin can you have a look?

File: doc/source/cuds_definition.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  **Figure. 2:** UML description of a CUDS Containers
+
+The CUDS containers, with the exception of the Lattice, are multi-type
+container (i.e. contain different types of objects). For each type the
+following operations are currently supported:
+
+* Add an item to the container
+* Get an item from the container based on the ``uid``
+* Remove an item from the container based on the ``uid``
```
Comment (roigcarlo) [inline doc/source/cuds_definition.rst]: Actually, mesh elements / points do not support to be removed from a mesh, as this operation was not reflected in the wiki:

https://publicwiki-01.fraunhofer.de/SimPhoNy-Project/index.php/Mesh_description
https://publicwiki-01.fraunhofer.de/SimPhoNy-Project/index.php/Mesh_data

It shouldn't be a problem to implement but I ignore if the lack of this feature is intended or not.

Comment (itziakos) [inline doc/source/cuds_definition.rst]: hm.. that is correct. I could think that we forgot about it. I will correct the documentation and raise an API change on the mesh workgroup

File: doc/source/cuds_definition.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+.. note::
+
+   For Lattice items do not have a ``uid`` but are accessed based on
+   ``(i,j,k)`` index tuples.
+
+.. rubric:: Snapshot principle
+
+All CUDS containers (native or proxy based) are owners of their data
+and will always return cuds components that container copy (snapshot)
```
Comment (nathanfranklin) [inline doc/source/cuds_definition.rst]: cuds -> CUDS

"components that container a copy (snapshot)"-> "components that contain a copy (snapshot)"

File: doc/source/cuds_definition.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+of the internal data representation. As a result running comparing the
```
Comment (nathanfranklin) [inline doc/source/cuds_definition.rst]: Suggestion:  The "As a result.." sentence could be changed to  "As a result, the information of an item extracted with a ``get`` operation before and and after a simulation run is not expected to be the same."

**Comment by nathanfranklin** [comment]
@itziakos ,  i like the descriptions in CUDS.   looks good; i just left a few suggestions for the text.

File: doc/source/cuds_definition.rst (hunk 1 above)
Comment (itziakos) [inline doc/source/cuds_definition.rst]: I have updated the uml diagram of the Mesh and added more info in the exceptions note

**Comment by itziakos** [comment]
@nathanfranklin and @roigcarlo  I have updated the PR, please have another look

**Comment by itziakos** [comment]
One thing that is bad with this PR is that it add the plantuml.jar binary file which is not nice and problematic for versioned controlled repositories

**Comment by nathanfranklin** [comment]
:+1:

**Comment by itziakos** [comment]
@nathanfranklin and @roigcarlo  I have removed the planuml.jar and fixed the mesh api description. Please have another look

**Comment by nathanfranklin** [comment]
:+1:

**Pull request merged**